tests/assets/test_cli_hint_suite.txt

```diff
@@ -74,18 +79,26 @@
 "SET key value KEEPTTL " "[NX|XX] [GET]"
 "SET key value XX " "[GET] [EX seconds|PX milliseconds|EXAT unix-time-seconds|PXAT unix-time-milliseconds|KEEPTTL]"
 
+# If an input word can't be matched, stop hinting.
+"SET key value FOOBAR " ""
+
 # Reordering partially-matched optional argument: GEORADIUS key longitude latitude radius M|KM|FT|MI [WITHCOORD] [WITHDIST] [WITHHASH] [COUNT count [ANY]] [ASC|DESC] [STORE key] [STOREDIST key]
 "GEORADIUS key " "longitude latitude radius M|KM|FT|MI [WITHCOORD] [WITHDIST] [WITHHASH] [COUNT count [ANY]] [ASC|DESC] [STORE key] [STOREDIST key]"
 "GEORADIUS key 1 2 3 M " "[WITHCOORD] [WITHDIST] [WITHHASH] [COUNT count [ANY]] [ASC|DESC] [STORE key] [STOREDIST key]"
 "GEORADIUS key 1 2 3 M COUNT " "count [ANY] [WITHCOORD] [WITHDIST] [WITHHASH] [ASC|DESC] [STORE key] [STOREDIST key]"
 "GEORADIUS key 1 2 3 M COUNT 12 " "[ANY] [WITHCOORD] [WITHDIST] [WITHHASH] [ASC|DESC] [STORE key] [STOREDIST key]"
-# The next case is not working yet - it needs to be fixed:
+"GEORADIUS key 1 2 3 M COUNT 12 " "[ANY] [WITHCOORD] [WITHDIST] [WITHHASH] [ASC|DESC] [STORE key] [STOREDIST key]"
+"GEORADIUS key 1 -2.345 3 M COUNT 12 " "[ANY] [WITHCOORD] [WITHDIST] [WITHHASH] [ASC|DESC] [STORE key] [STOREDIST key]"" ""
+# Wrong data type: latitude must be a double. Hinting fails.
+"GEORADIUS key 1 X " ""
+# TODO: The next case is not working yet - it needs to be fixed:
 # "GEORADIUS key 1 2 3 M COUNT 12 ASC " "[WITHCOORD] [WITHDIST] [WITHHASH] [STORE key] [STOREDIST key]"
```

Comment:
what's the problem here?
this is the test that proves `ANY` isn't hinted anymore, and it seems to work on your last version

Reply:
Simple - the problem is that I forgot to uncomment the test case... Yes, it works.

Reply:
> the one thing i noticed and not sure about is that when redis-cli works with an old server, we still have the old matching logic.
> i suppose that in order to improve that, we'll have to bundle something other than the plain old help.h.
> not sure we wanna proceed in that direction

The way I see it, there are a few options:

1. With an old server, fall back to the old matching logic (this is what the code currently does)
2. With an old server, disable arg matching entirely; just give an initial hint with the entire help string and then stop hinting.
3. With an old server. use some heuristics to translate the static syntax strings into a guessed arg spec structure, and use that to support the new matching logic. This might not be too bad, though it wouldn't be that great either.
4. Somehow use the command spec tables in `commands.c` as a fallback source for CLI hinting. Currently that's not practical since `commands.c` includes the `server.h` header file, and we don't want the CLI to do that. But maybe we could solve that somehow. We could even use the `since` fields to avoid hinting unsupported commands/args. This wouldn't give hinting for module commands, though.

Thoughts?

@oranagra

Reply:
a variant of 4 would be to embed a cached copy of commands.json inside the cli (instead of the help.h we have now).
then when COMMAND DOCS fails, we issue an INFO SERVER instead and use the `since` fields in our cached json to do filtering.
maybe in some way it would be easier to store the RESP response of COMMAND DOCS in some way, instead of parsing a json.

anyway, it looks like a big investment, so we have to consider the pros and cons, since arguably we can maybe drop the hinting entirely (2).

what i don't like about the current approach (1) is that it looks like a bug, and also that there's probably quite a lot of code to back it up (including help.h).

so arguably, if we invest some effort in either parsing a json, or compiling commands.c into the cli, we'll still be deleting a lot more code that we're adding (since the majority of the code is shared with the case COMMAND DOCS does exits).

Reply:
The CLI currently has code to parse the RESP response into an internal structure. It would be annoying to have to add parallel code to do the same for JSON. If we could store the RESP, all that would be missing would be the version filtering. (Plus, of course, we still wouldn't have filtering for module commands.)

Reply:
It parses RESP by using hiredis, it's not the same as handing it a blob of bytes.

Reply:
What I meant is that there is code to read the redis reply structure into the command args table. A similar chunk of code would be needed to consume JSON instead (actually, in addition).

But yeah, we'd also need the RESP parser.

Reply:
Looking some more at commands.c, I'm not that happy about compiling it into the cli. The command table currently combines the command arg specs and docs with pointers to the actual implementation functions for the commands. It's both a syntax table and a dispatch table, and probably some other things too. Integrating it into the cli code exposes the client to stuff that rightfully should stay server-side.

If we want to go that route (of precompiling the arg specs table) I think the "right" way to do it is to autogenerate a separate command table for the CLI. Basically, a replacement for help.h that would fill in the whole `commandDocs` table, including all the arg specs. It would not be necessary to parse anything at runtime, or to have access to the JSON or RESP specs.But it's a lot of data to keep around just for backwards-compatibility - commands.c is 7,400 lines long, and it would presumably be similar. help.h is

Reply:
(The *.json files are 18,000 lines!)

Reply:
Okay, I've spent some more time on the issue of legacy servers.

A workable solution involves the following:
1. Generating a new `cli-commands.c` file containing C structs with only the client-side part of the argument specs. This comes to about 4500 lines of generated code, plus 250 lines of Python to generate it.
2. Adding code to populate the help tables from that data if `COMMAND DOCS` is not available, filtering by `since`. That adds about 50 lines of code to redis-cli.c, I think.

This would give us full hinting support for legacy servers, except for module commands. Is it worth it? Should I finish the implementation?

@oranagra

Reply:
if only we could rely on having python available at compile time, then we wouldn't have to maintain these 4500 lines in the repo.
maybe we can change the existing python script to add some preprocessor macros to commands.c so that when included from redis-cli.c it'll not include the parts we don't want?

this would mean that the current commands.c will be slightly bigger, but we won't have another 4500+250 extra lines sitting in the repo.

Reply:
Bear in mind that this would replace `help.h` (1860 lines) and `generate-command-help.rb` (150 lines), so at least there's that.

I don't think macros would solve this cleanly, since there are a lot of struct fields in commands.c we don't need/want, in particular the function pointers. Maybe if we `#define` the function names to NULL... it starts to get very ugly.

Reply:
that seems easy to resolve.
instead of generating lines like this:
```c
{"bitcount","Count set bits in a string","O(N)","2.6.0",CMD_DOC_NONE,NULL,NULL,COMMAND_GROUP_BITMAP,BITCOUNT_History,BITCOUNT_tips,bitcountCommand,-2,CMD_READONLY,ACL_CATEGORY_BITMAP,{{NULL,CMD_KEY_RO|CMD_KEY_ACCESS,KSPEC_BS_INDEX,.bs.index={1},KSPEC_FK_RANGE,.fk.range={0,1,0}}},.args=BITCOUNT_Args},
```

it can wrap the reference to the command proc with `CMD_PROC()` macro.
```c
{"bitcount","Count set bits in a string","O(N)","2.6.0",CMD_DOC_NONE,NULL,NULL,COMMAND_GROUP_BITMAP,BITCOUNT_History,BITCOUNT_tips,CMD_PROC(bitcountCommand),-2,CMD_READONLY,ACL_CATEGORY_BITMAP,{{NULL,CMD_KEY_RO|CMD_KEY_ACCESS,KSPEC_BS_INDEX,.bs.index={1},KSPEC_FK_RANGE,.fk.range={0,1,0}}},.args=BITCOUNT_Args},
```

then, it can do this on the top:
```c
#ifndef CMD_PROC
#define CMD_PROC(proc) proc
#endif
```
all that's left is for redis-cli.c to declare `#define CMD_PROC(proc) NULL` before including commands.c.

Reply:
Hmm... that could work... I'll take another look. (Though again, there's a lot more "stuff" on that line not relevant to the cli. Preprocessing it all away won't be pretty. I'm not sure it's better than maintaining two parallel generated files that each contain only what they're supposed to have.)

Reply:
similar thing can be done with other pieces of information that we don't want. like extracting the key-specs to an external object, like we do for args.

Reply:
Okay, I tried extracting key-specs to external tables, and everything compiles and the tests pass... and hinting works for legacy servers.

**Except** it seems to have broken the module API. (See latest tests.) I don't know the module code well enough to understand how I broke it. All I did was change `key_specs_static` from an embedded array in the command struct to a pointer to an external array.

Reply:
Never mind - I fixed it. Expecting all tests to pass now.